llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -26057,6 +26059,38 @@ performScalarToVectorCombine(SDNode *N, TargetLowering::DAGCombinerInfo &DCI,
   return NVCAST;
 }
 
+static SDValue performSHLCombine(SDNode *N, SelectionDAG &DAG) {
+  SDValue Op0 = N->getOperand(0);
+  SDValue Op1 = N->getOperand(1);
+  EVT VT = N->getValueType(0);
+  if (VT != MVT::i32 && VT != MVT::i64)
+    return SDValue();
+
+  // If the operand is a bitwise AND with a constant RHS, and the shift is the
+  // only use, we can pull it out of the shift.
+  //
+  // (shl (and X, C1), C2) -> (and (shl X, C2), (shl C1, C2))
+  if (!Op0.hasOneUse() || Op0.getOpcode() != ISD::AND)
+    return SDValue();
+
+  ConstantSDNode *C1 = dyn_cast<ConstantSDNode>(Op0.getOperand(1));
+  ConstantSDNode *C2 = dyn_cast<ConstantSDNode>(Op1);
+  if (!C1 || !C2)
+    return SDValue();
+
+  // Might be folded into shifted op, do not lower.
+  unsigned UseOpc = N->use_begin()->getOpcode();
+  if (N->hasOneUse() &&
+      (UseOpc == ISD::ADD || UseOpc == ISD::SUB || UseOpc == ISD::SETCC ||
+       UseOpc == AArch64ISD::ADDS || UseOpc == AArch64ISD::SUBS))
+    return SDValue();
+
+  SDLoc DL(N);
+  SDValue NewRHS = DAG.getNode(ISD::SHL, DL, VT, Op0.getOperand(1), Op1);
```

Comment:
I think it's a bit clearer to use `C1` and `C2` here instead of `Op0.getOperand(1)` and `Op1`, because matches the example combine in your comment above.